### ISSUE TITLE
Driverkit fails to build drivers for COS kernel 6.1.58+

### ISSUE THREAD

**Issue by clr34m3r** [falcosecurity/driverkit#344]
**Describe the bug**

Problem building drivers for COS kernel 6.1.58+

**How to reproduce it**

I am encountering an issue when trying to upgrade my GKE to version 1.28 and compile the Falco eBPF probe.
Here is the relevant part of the Falco log that outlines the problem:
```
* Setting up /usr/src links from host
* Running falco-driver-loader for: falco version=1.8.0-611, driver version=5c0b863ddade7a45568c0ac97d037422c9efb750
* Running falco-driver-loader with: driver=bpf, compile=yes, download=yes
* Mounting debugfs
* Trying to download a prebuilt eBPF probe from https://download.falco.org/driver/5c0b863ddade7a45568c0ac97d037422c9efb750/falco_cos_6.1.58%2B_1.o
curl: (22) The requested URL returned error: 404
Unable to find a prebuilt falco eBPF probe
* COS detected (build 17800.66.78), using COS kernel headers
* Found kernel config at /proc/config.gz
* Downloading https://storage.googleapis.com/cos-tools/17800.66.78/kernel-headers.tgz
* Extracting kernel sources
* Configuring kernel
* Trying to compile the eBPF probe (falco_cos_6.1.58+_1.o)
warning: the compiler differs from the one used to build the kernel
  The kernel was built by: Chromium OS 16.0_pre484197_p20230405-r10 clang version 16.0.0 (/var/tmp/portage/sys-devel/llvm-16.0_pre484197_p20230405-r10/work/llvm-16.0_pre484197_p20230405/clang 2916b99182752b1aece8cc4479d8d6a20b5e02da)
  You are using:           gcc (Debian 5.5.0-12) 5.5.0 20171010
scripts/mod/modpost: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by scripts/mod/modpost)
scripts/mod/modpost: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by scripts/mod/modpost)
make[2]: *** [scripts/Makefile.modpost:126: /usr/src/falco-5c0b863ddade7a45568c0ac97d037422c9efb750/bpf/Module.symvers] Error 1
make[1]: *** [Makefile:1966: modpost] Error 2
make: *** [Makefile:18: all] Error 2
mv: cannot stat '/usr/src/falco-5c0b863ddade7a45568c0ac97d037422c9efb750/bpf/probe.o': No such file or directory
Unable to load the falco eBPF probe
Fri Apr 26 04:01:19 2024: Falco version 1.8.0-611 (driver version 5c0b863ddade7a45568c0ac97d037422c9efb750)
Fri Apr 26 04:01:19 2024: Falco initialized with configuration file /etc/falco/falco.yaml
Fri Apr 26 04:01:19 2024: Loading rules from file /etc/falco/falco_rules.yaml:
Fri Apr 26 04:01:19 2024: Loading rules from file /etc/falco/falco_rules.local.yaml:
Fri Apr 26 04:01:20 2024: Unable to load the driver.
Fri Apr 26 04:01:20 2024: Runtime error: can't open BPF probe '/root/.falco/falco-bpf.o': Errno 2. Exiting.
```

In an attempt to resolve this, I tried using Driverkit to build a suitable driver for Falco with the following config file content:
```
target: debian
driverversion: 5c0b863ddade7a45568c0ac97d037422c9efb750
kernelrelease: 6.1.58+
kernelversion: 1
output:
  module: /tmp/falco_cos_6.1.58%2B_1.ko
  probe:  /tmp/falco_cos_6.1.58%2B_1.o
```

However, I encountered validation errors and fatal execution errors with Driverkit:
```
2024-05-15 15:29:10 INFO  using config file file: cos.yaml
2024-05-15 15:29:10 ERROR error validating build options err: both module and probe are not supported by given options
2024-05-15 15:29:10 FATAL error executing driverkit err: exiting for validation errors
```

I tried to replace kernelversion with 6.1.58 and encountered new errors:
```
2024-05-15 15:57:07 INFO  using config file file: cos_1.yaml
2024-05-15 15:57:07 DEBUG running with options
                      ├ output-module: /tmp/falco_cos_6.1.58%2B_1.ko
                      ├ output-probe: /tmp/falco_cos_6.1.58%2B_1.o
                      ├ driverversion: 5c0b863ddade7a45568c0ac97d037422c9efb750
                      ├ kernelrelease: 6.1.58
                      ├ kernelversion: 1
                      ├ target: debian
                      ├ arch: amd64
                      ├ kernelurls: []
                      ├ repo-org: falcosecurity
                      └ repo-name: libs
2024-05-15 15:57:07 INFO  starting build processor: docker
2024-05-15 15:57:09 FATAL error executing driverkit err: kernel headers not found
```

Even though the Falco logs seem to suggest that it can find the kernel headers at [https://storage.googleapis.com/cos-tools/17800.66.78/kernel-headers.tgz](https://storage.googleapis.com/cos-tools/17800.66.78/kernel-headers.tgz), driverkit refuses to found it.

I'm looking for guidance and assistance on how to resolve this issue with Driverkit. Any help or insights from the community would be greatly appreciated.

Thank you in advance!

**Environment**

- OS: Container-Optimized OS
- Kernel: [COS-6.1.58](https://cloud.google.com/container-optimized-os/docs/release-notes/m109#cos-109-17800-66-78_)

**Comment by FedeDP** [comment]
Hi! Thanks for opening this issue.

> scripts/mod/modpost: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.33' not found (required by scripts/mod/modpost)
scripts/mod/modpost: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.34' not found (required by scripts/mod/modpost)


This comes from falco-driver-loader; which version are you using?  Latest falco-driver-loader ships with up to date GLIBC version and should work fine:
```
/usr/lib/x86_64-linux-gnu/libc.so.6
GNU C Library (Debian GLIBC 2.36-9+deb12u4) stable release version 2.36.
```

> In an attempt to resolve this, I tried using Driverkit to build a suitable driver for Falco with the following config file content:

Driverkit does support cos as vanilla, ie: you must give it the kernelrelease (6.1.58 in your case) and a full kernel config base64 encoded. But cos needs some more logic than that, that's why falco-driver-loader (nowadays falcoctl driver-loader) has specific logic for it: https://github.com/falcosecurity/falcoctl/blob/main/pkg/driver/distro/cos.go

To summarize: driverkit does not support cos out of the box :) i think that if you use `falco-driver-loader:latest` (ie: 0.37.1) you should be fine.
